xmtp_mls/README.md

```diff
@@ -0,0 +1,143 @@
+# XMTP MLS
+
+## Database Schema
+
+Foreign key constraints and indexes omitted for simplicity.
+
+```sql
+-- Would be a table that only holds a single identity (your own)
+CREATE TABLE identities {
+    id TEXT PRIMARY KEY NOT NULL, -- The installation_id of the identity
+    signature_keypair BLOB NOT NULL,
+    "credential" BLOB NOT NULL,
+    created_at_ns BIGINT NOT NULL,
+    wallet_address TEXT NOT NULL,
+}
+
+-- Minimum viable group schema
+CREATE TABLE groups {
+    id BLOB PRIMARY KEY NOT NULL, -- Random ID generated by group creator
+    created_at_ns BIGINT NOT NULL, -- Based on the timestamp of the welcome message
+    membership_state INT NOT NULL, -- Enum of the GROUP_MEMBERSHIP_STATE listed below
+}
+
+-- Successfully processed messages meant to be returned to the user
+CREATE TABLE group_messages {
+    id BLOB PRIMARY KEY NOT NULL -- Derived via SHA256(CONCAT(decrypted_message_bytes, conversation_id, timestamp))
+    group_id BLOB NOT NULL,
+    decrypted_message_bytes BLOB NOT NULL, -- Message contents after decryption
+    sent_at_ns BIGINT NOT NULL, -- Based on the timestamp of the message
+    sender_installation_id TEXT NOT NULL,
+    sender_wallet_address TEXT NOT NULL, -- Could remove this if we added a table mapping installation_ids to wallet addresses
+}
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state {
+    topic TEXT PRIMARY KEY NOT NULL,
+    last_message_timestamp BIGINT NOT NULL
+}
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents {
+    id SERIAL PRIMARY KEY NOT NULL,
+    kind INT NOT NULL, -- INTENT_KIND
+    group_id BLOB NOT NULL,
+    "data" BLOB NOT NULL, -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "state" INT NOT NULL, -- INTENT_STATE,
+    message_hash BLOB -- The hash of the encrypted, concrete, form of the message if it was published.
+}
+
+CREATE TABLE outbound_welcome_messages {
+    id BLOB PRIMARY KEY NOT NULL, -- Derived via SHA256(CONCAT(group_id, welcome_message, ratchet_tree, installation_id))
+    state INT NOT NULL, -- OUTBOUND_WELCOME_STATE
+    installation_id BLOB PRIMARY KEY NOT NULL,
```

Comment:
Curious what the installation_id here is referring to? I don't think it should be an individual recipient, as Welcome messages are shared by all recipients

Reply:
An `installation_id` is the term we are using as the unique identifier for particular app installation. It is just the public key of the signing key for their MLS identity.

While welcome messages are shared between all recipients, we are going to need to send multiple copies of the message to each installation's topic on our network so that they can go to the correct recipient. Every installation has a topic where their welcome messages are stored.

It's a little wasteful to keep so many copies of each pending message, but it lets us modify the state if some sends succeed and some fail.